### PR TITLE
feat: add replaced by events to commit page and backend

### PR DESCRIPTION
**Pull request by miguel-crespo-fdc**
When some release is replaced by another on some environment, a new "Replaced by" event is registered and associated with the release. There is a link on the release page showing the other release that it was replaced by.

Screenshot 2024-03-12 at 14 35 00

### CODE REVIEW

File: services/cd-service/pkg/repository/transformer_test.go [hunk 1]
```diff
@@ -1675,6 +1675,149 @@ func TestApplicationDeploymentEvent(t *testing.T) {
...
+			if verErr != nil {
+				if !(strings.Contains(verErr.Error(), tc.ExpectedError)) {
+					t.Fatalf("want :\n\"%v\"\nbut got:\n\"%v\"", tc.ExpectedError, verErr.Error())
+				}
+				if tc.ExpectedError == "" {
+					t.Fatalf("Error while verifying content: %v. Filesystem content:\n%s", verErr, strings.Join(listFiles(fs), "\n"))
+				}
+
+			} else if tc.ExpectedError != "" {
+				t.Fatalf("Expected error %s, but got noting.", tc.ExpectedError)
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/transformer_test.go:1815]:
```suggestion
				t.Fatalf("Expected error %s, but got nothing.", tc.ExpectedError)
```

File: services/cd-service/pkg/repository/transformer_test.go [hunk 2]
```diff
@@ -1675,6 +1675,149 @@ func TestApplicationDeploymentEvent(t *testing.T) {
...
+
+			if verErr != nil {
+				if !(strings.Contains(verErr.Error(), tc.ExpectedError)) {
+					t.Fatalf("want :\n\"%v\"\nbut got:\n\"%v\"", tc.ExpectedError, verErr.Error())
+				}
+				if tc.ExpectedError == "" {
+					t.Fatalf("Error while verifying content: %v. Filesystem content:\n%s", verErr, strings.Join(listFiles(fs), "\n"))
+				}
+
+			} else if tc.ExpectedError != "" {
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/transformer_test.go:1814]: Why the `else`?
```suggestion
			}
			if tc.ExpectedError != "" {
```

**Review by sven-urbanski-freiheit-com — CHANGES_REQUESTED** [review]
(see other comments)

File: services/cd-service/pkg/repository/transformer.go [hunk 3]
```diff
@@ -1692,36 +1707,60 @@ func (c *DeployApplicationVersion) Transform(
...
+			logger.FromContext(ctx).Sugar().Infof(
+				"The source commit ID %s is not a valid/complete SHA1 hash, event cannot be stored.",
+				commitID)
+			return nil
+		}
+		if err := writeEvent(eventUuid, commitID, fs, ev); err != nil {
+			return fmt.Errorf(
+				"could not write an event for commit %s, error: %w",
+				commitID, err)
+		}
```
[the remark this answers is not in the document]
Comment (miguel-crespo-fdc) [reply services/cd-service/pkg/repository/transformer.go]: Can't simply return result of the writeEvent function, breaks idempontency test

[the remark this answers is not in the document]
Comment (sven-urbanski-freiheit-com) [reply services/cd-service/pkg/repository/transformer.go]: ok, let's leave it as is for now.
We'll tackle this here SRX-UG7OJV

**Review by sven-urbanski-freiheit-com — APPROVED** [review]

**Pull request merged**